### ISSUE TITLE
Apply Retry attribute to all methods in fixture or assembly

### ISSUE THREAD

**Issue by MetaSam** [nunit/nunit#2452]
In NUnit 2, we had a Retry attribute in place that applied the Retry logic to every method in an assembly. My understanding is that in NUnit 3, it is not possible to apply a behavior to all methods of a fixture or assembly via an attribute with the current extensibility options (correct me if I'm wrong!).

Is there any way that we can enable this behavior with extensions, or is there any chance that at some point the Retry attribute can be applied to fixture and assembly? I think it could also be very useful for the MaxTime attribute.

Note: I know that in general using the Retry attribute is not good practice, but despite our best efforts to fix this, our Selenium test suite still has many flaky tests and the Retry solution with NUnit 2 worked well for us.

**Comment by ocmyhome** [comment]
This is a feature we'd like to have as well (Retry on Fixture or Assembly level).  We have about 70+ e2e tests and growing.  Many them however are randomly flaky.  Setting [Retry] on each test method is a bit tedious.  Would love to have this feature worked on...

**Comment by CharliePoole** [comment]
The feature seems somewhat desirable. It could be done on an adhoc basis - i.e. allow `RetryAttribute` at higher levels than the method and apply it to all methods. I'd be opposed to that.

The better approach, which is in fact referenced in this issue's description, is to establish a standard way to create built-in or custom attributes that apply to all tests in a given scope: Assembly, Namespace or Fixture. Once we have such an interface or other API, then this could be implemented using it. It would be up to the @nunit/framework-team to decide whether this particular request warrants inclusion directly in the framework or if it should be left up to custom attributes created by users.

@jnm2 and I have had some discussion about the best way to implement such an interface in issue #2666.

**Comment by rprouse** [comment]
I think this is one of those features that should be left up to users to create as a custom attribute. There many possible ways a higher level Retry attribute could work. Does it keep retrying the fixtures? Does it apply retry to the methods? How does it work in combination with `Order`?

I also see Retry as a bit of an ant-pattern. I understand why people want it and am sympathetic, but, like `ExpectedException`, I think we need to draw a line somewhere.

So, I agree with @CharliePoole that we should ensure that our extensibility allows users to create these attributes and we could even provide a simple example, like we did with `ExpectedException` in the C# samples project.

**Comment by ocmyhome** [comment]
I'm not a developer, so any examples would be much appreciated.

**Comment by mcgary-andrew** [comment]
I am not sure if this is something that I should put here or in a new issue, but I am adding the Retry attribute on some of my tests but when I run a group of tests through VS Test Explorer, the failing tests are not being retried. If I just run that 1 single test by itself, it will retry if it fails.

**Comment by CharliePoole** [comment]
Your problem is likely to be lost if it's appended on a different issue, so you should file it separately. Before doing that, you should ascertain if the problem is exclusive to using the test adapter and - if it is - file the problem under the nunit vs test adapter.

**Comment by LirazShay** [comment]
@CharliePoole 
I've created custom attribute (copied from RetryAttribute but added AttributeTargets.Assembly)  
It doesn't retry when attribute was added to Assembly, though it works not method level
[MyRetryAttribute.cs.txt](https://github.com/nunit/nunit/files/2280813/MyRetryAttribute.cs.txt)
Am I missing somthing?  
How can I implement it by extension or any other workaround? 
I'm using SpecFlow and all the nunit tests are auto-generated, so I can only add assembly level attributes

**Comment by CharliePoole** [comment]
Two problems here...

1. If your attribute worked, it would run the entire assembly multiple times on failure. That's what retry attribute does to the test it is placed on. And an assembly is a test, just like a test case or fixture.

2. The attribute doesn't work anyway because the commands used internally for tests case and test suite execution are entirely different. That is, in my view, a deficiency in NUnit that somebody might correct in the future, but it's how it works right now.

An attribute that does modifies a test is quite different from one that stores information to be used to modify all child tests below that point. RetryAttribute is an example of the first. TimeoutAttribute is a good example of an attribute that actually does both - it operates in one way when placed on a test case and in another way when placed on a fixture.

My guess is that this problem is easier to resolve in the SpecFlow code generation than in NUnit itself.

**Comment by akamud** [comment]
I'd love to have this feature for UI tests as well. Retrying is a common necessity for these kinds of tests, having a way of specifying this only once per Test Suite would be very useful.

**Comment by CharliePoole** [comment]
@nunit/framework-team This issue has been here for two years now. It has neither been accepted as a potential feature nor rejected, at least formally. There is some indication in the comments about how we wanted to handle it, however.

About four months into the two years, @rprouse and I gave opinions of what should happen. We talked about inventing some standard way to tell NUnit to apply a particular attribute that normally only works on methods more generally, i.e. to all the test methods in a fixture.

Somebody would have to design an approach and somebody would have to implement it. As a start, we could create an issue to do it and finally close this one.

Can we reach some team consensus as to how to handle this issue? I'll be glad to draft a new issue, if that would help.

**Comment by rajatjangir** [comment]
+1 for adding this feature

**Comment by akytsun** [comment]
+1 too

**Comment by Simford** [comment]
+1

**Comment by pavliukStanislav** [comment]
+1

**Comment by vitali-sonchyk-epam** [comment]
+1

**Comment by Yalkl** [comment]
+1

**Comment by alexander-kucherov** [comment]
+1

**Comment by akshaysarode31** [comment]
+1 
is someone working on this feature?

**Comment by Toumash** [comment]
@akshaysarode31 no, I'm using a wrapper with retry inside all my tests. My scenario was using nunit to run e2e tests, so there was 
```
await page.Run(async ()=> { /* some code */});
```

so that scenario was easy to add a try - i've just created an extension method
```
await page.Execute(async () => {/* some code */});
```

as 
```
   public static async Task Execute(this IPage page, Func<Task> action, [CallerMemberName] string callerName = "")
    {
        var testRunCount = 0;
        var success = false;
        do
        {
            try
            {
                testRunCount++;
                await action();
                success = true;
            }
            catch (Exception e)
            {
                var thatWasTheLastRetry = testRunCount == AppSettingsProvider.RetryCount();
                if (thatWasTheLastRetry)
                {
                    await HandleException(page, callerName, e, testRunCount);
                    throw;
                }
            }
        } while (!success && testRunCount <= AppSettingsProvider.RetryCount());
    }
```

**Comment by prc0mputing** [comment]
I solved this issue by developing a new retry attribute. You can view it in my repo [here](https://github.com/prc0mputing/nunit-retry-attribute).

It is not easy to implement a decorator (GOF) in NUnit3. So I used a trick that I discovered in the NUnit3 repo unit test, where the test method is wrapped by a method wrapper class. This attribute adds required attributes over the underlying test cases of an assembly or test fixture class.

**Comment by akshaysarode31** [comment]
@prc0mputing Thank you for looking into this. I Will definitely try your solution

**Comment by marco-eckstein** [comment]
@prc0mputing Thanks! Your code was helpful in developing a similar solution for a project I am working on.

**Comment by OsirisTerje** [comment]
@nima-bard Have you considered packaging up your [nunit-retry-attribute](https://github.com/nima-bard/nunit-retry-attribute/tree/main) solution and release it on nuget ?  That would make it more visible, and we could add some info on the nunit docs about it.

**Comment by davidvesely** [comment]
@nima-bard your solution works perfectly. Just one note for anyone who tries this approach - if applying the custom attribute on class or assembly level isn't working, make sure that `[TestFixture]` attribute is presented on the test class. Otherwise NUnit context is not reading properly the test methods and `ITest.Tests` collection will be empty.

**Comment by CharliePoole** [comment]
@davidvesely 

If this is the case, it sounds like a bug, either in NUnit or in the custom attribute.

`TestFixtureAttribute` is needed only if you want to supply constructor arguments to a parameterized fixture. But the bare `[TestFixture]` has been optional for many years, provided the class contains tests. A custom retry attribute should not change that.
